### ISSUE TITLE
Plans for Expanding the DDCR Service

### ISSUE THREAD

**Issue by karafecho** [ResearchSoftwareInstitute/greendatatranslator#146]
The purpose of this issue is to track potential plans for expanding the DDCR Service.

**Comment by karafecho** [comment]
Current plans:
1. Expand existing service by:
- Adding additional feature variables to explore questions of relevance to FA (metformin, WBC, hematocrit, platelets) (Maureen Hoatlin)
- Adding additional variables of relevance to asthma (e.g., pulmonary function tests) (Dave Peden)
- Creating new tables for years 2012 - present (or 2012 - 2017)
2. NIEHS EPR Collaboration and asthma demonstration use case (Charles Schmitt)
3. Create new cohorts:
- Diabetes (Miriam Udler)
- Liver-induced drug injury (Paul Watkins)

**Comment by xu-hao** [comment]
Dependencies: 

See @karafecho's post on current plans

**Comment by karafecho** [comment]
Also see #122.

**Comment by karafecho** [comment]
09/06/18: Received approval from CDWH Oversight Committee to: (1) work with UNC's IRB and the CDWH Operations Committee on any plans to develop new ICEES cohorts; (2) disseminate ICEES beyond the Translator program; and (3) allow a variety of programming techniques to query the ICEES API.

**Comment by karafecho** [comment]
Obtained additional years CDWH data: 2012, 2013, 2015, 2016. Obtained nationwide CMAQ PM2.5 and ozone estimates for 2002-2014. Obtained nationwide ACS estimates for most recent survey period (2012-2016). Hao to integrate using FHIR pipeline. ETA 10/13/18.

**Comment by karafecho** [comment]
Update on where we are, in terms of integrated ICEES tables:

- 2010, 2011 ICEES tables (patient and visit) generated using the manual pipeline and including CDWH patient data, 2012-2016 ACS data (one dataset from 5-year survey), 2010 & 2011 CMAQ data on PM2.5 and ozone from first batch (large grid sizes), 2016 TIGERline roadway data, and 2016 HPMS roadway data.
- 2010 ICEES tables (patient and visit) generated using CAMP FHIR/FHIR PIT and including i2b2 -> FHIR patient data with expanded fields/variables, 2012-2016 ACS data (one dataset from 5-year survey), 2010 & 2011 CMAQ data on PM2.5 and ozone from first batch (large grid sizes), 2016 TIGERline roadway data, and 2016 HPMS roadway data

We also have new (or relatively new) data that have not yet been integrated with patient data:

- 2002-2015 CMAQ data on PM2.5, ozone, acetaldehyde, CO, NO, NO2, NOx, and SO2, plus benzene for the years 2002, 2006, 2013, 2014 (12-km grid size)
- 2007-2011 and 2012-2016 ACS estimates plus standard errors (one dataset from each 5-year survey sample)
- 2015 CAMP FHIR i2b2 -> FHIR patient data with expanded fields/variables

Our goal is to create new integrated ICEES tables for 2011-2016 (patient and visit) using CAMP FHIR/FHIR PIT and including the new data listed above, plus the older data, mapped appropriately to the relevant year of patient data. I believe we decided to use PCORnet as the source data for CAMP FHIR, so we may want to create a third table for 2010, just for consistency across years. @empff: please confirm the use of PCORnet for the new tables, 2010-2016.

**Comment by karafecho** [comment]
Update:

CAMP FHIR-generated i2b2 -> FHIR patient/visit files for 2010 were used for validation testing

CAMP FHIR-generated PCORnet -> FHIR patient/visit files for 2010 and 2015 were created and sent to Hao for integration using FHIR PIT

CAMP FHIR-generated PCORnet -> FHIR patient/visit files for 2011 and 2016 are nearly complete

CAMP FHIR-generated PCORnet -> FHIR patient/visit files for 2012, 2013, and 2014 will take a bit longer to complete

**Comment by karafecho** [comment]
Corrections and update:

CAMP FHIR-generated i2b2 -> FHIR patient/visit files for 2010 were used for validation testing

CAMP FHIR-generated i2b2 -> FHIR patient/visit files for 2011 were sent to Hao for integration using FHIR PIT

CAMP FHIR-generated i2b2 -> FHIR patient/visit files for 2012-2014 are forthcoming

CAMP FHIR-generated PCORnet -> FHIR patient/visit files for 2015 were sent to Hao for integration using FHIR PIT

CAMP FHIR-generated PCORnet -> FHIR patient/visit files for 2016 are nearly complete

**Comment by karafecho** [comment]
From Hao re LOINC2HPO tool:

I've talked to the developer of LOINC2HPO. And it seems like it is not too hard to integrate. But my impression from the conversation was that that we need to refactor their code into a library in order to use it.

@xu-hao : is this something that a student intern may be able to help with?

**Comment by karafecho** [comment]
Per James, 07.29.19:

I loaded the 2016 data on to Rockfish in folder /home/champioj/FHIR/. It's names 2016.tar.

**Comment by karafecho** [comment]
@jameschump : Thanks for loading the 2016 data. Any updates on the 2012, 2013, 2014 data? ETA? Part of the hackathon plan is to inventory Translator Clinical Data Sources, so I'm hoping we can provide a "complete" set of tables, if that makes sense.

**Comment by empfff** [comment]
@karafecho funny you should ask, we should have it all by next week. 2012 is just about ready to go, 2013 and 2014 to follow shortly thereafter.

**Comment by jameschump** [comment]
> 
> 
> @jameschump : Thanks for loading the 2016 data. Any updates on the 2012, 2013, 2014 data? ETA? Part of the hackathon plan is to inventory Translator Clinical Data Sources, so I'm hoping we can provide a "complete" set of tables, if that makes sense.

I will get to it this afternoon.

**Comment by karafecho** [comment]
@empfff @jameschump : Great! Thanks.

**Comment by jameschump** [comment]
@karafecho @xu-hao @empfff Okay, 2012 is out there in /opt/RENCI/output/FHIR/2012

**Comment by xu-hao** [comment]
Thanks!
________________________________

**Comment by xu-hao** [comment]
2012, 2013 : integrated with env data, will do acs and nearest road next
2014: received data
2010, 2015: waiting for issues to be resolved upstream

**Comment by karafecho** [comment]
Thanks for the update, Hao.

@jameschump : where do we stand with the 2011 and 2016 patient data? also, what have the issues been with the 2010 and 2015 data?

**Comment by karafecho** [comment]
@jameschump @xu-hao @empfff @stevencox : Just so you all are aware, Stan and I are heading to the CDC on November 1 (exactly one month from now). The purpose of this visit is to talk about ICEES, the exposures data, and open science. NCATS will be there, including Noel (on behalf of Christine) and possibly Chris Austin (schedule permitting). It would be really great to have the new ICEES tables loaded behind the API and accessible prior to the meeting. Any chance we can make this a reality?

**Comment by karafecho** [comment]
@jameschump @xu-hao @empfff @stevencox : Any updates, folks?

**Comment by empfff** [comment]
i believe @jameschump and i have provisioned all data to rockfish for all years as of a few weeks ago. if this is not the case, please let us know.

**Comment by karafecho** [comment]
@jameschump : My understanding was that there were issues with some of the data that you provided and that you were working with Hao to address the issues. Has everything been addressed? @xu-hao : Do you have everything you need?

**Comment by jameschump** [comment]
Hao noticed a possible issue with the encoding type which I am currently looking into. Had to do with some of the Greek letters in the medication files.

**Comment by karafecho** [comment]
Thanks, James. We need to get rolling on this, so perhaps you can provide an ETA?

**Comment by jameschump** [comment]
I'm working on it now and plan to have it resolved today.

**Comment by xu-hao** [comment]
James is working on it. I don't need anything else for now.
________________________________

**Comment by xu-hao** [comment]
@jameschump I just remembered another question is that in 2015 lab some has effectiveDateTime field some has issued field.
________________________________

**Comment by jameschump** [comment]
Spoke with Hao this morning regarding the confusion around effectiveDateTime and issued fields and explained how they are populated from pcornet database vs i2b2. Also discussed the character encoding issue and what I need to do to fix it which I am working on now and hope to have corrected by the end of today.

**Comment by karafecho** [comment]
Great! Thanks for the hard work, James and Hao.

**Comment by jameschump** [comment]
I handed off an updated file with the correct character encoding so it should work now. Hao is now running through his pipeline.

**Comment by karafecho** [comment]
Update, 10.23.19: Hao is reviewing the 2010 to 2016 clinical data. He identified issues with several years of data (2010, 2011, 2014, 2015, 2016) and working with James to resolve them. Most of the issues have been resolved, although some remain. The 2012 and 2013 clinical data have been run through FHIR PIT and are now integrated with the environmental exposures data.

In an effort to prioritize the interests of folks at EPR, Hao will focus on extending FHIR PIT to accommodate the EPR data and run an additional integration with the 2012 and 2013 integrated data sets. He will then run the additional years of clinical data through FHIR PIT, including the EPR extension of the pipeline.

**Comment by karafecho** [comment]
Update, 10.25.19: PCORnet meds from 2010 are unavailable; i2b2 meds from 2010 will be used as a substitute.

**Comment by karafecho** [comment]
@jameschump @empfff @xu-hao : With the Translator Phase II proposals successfully submitted, I thought I'd check on the status of the new tables. Any updates?

**Comment by xu-hao** [comment]
I've finished generating unbinned new table for 2012 (including acs, ur, nearest road data, previous update only include the env data). 

I encountered some issues with 2013 and @jameschump has helped resolve them and I'm currently working on it.

**Comment by karafecho** [comment]
Great! Thanks, Hao.

@jameschump : Where do things stand with the patient data for years 2010, 2011, 2014, 2015, and 2016?

**Comment by karafecho** [comment]
@xu-hao : Just to clarify, the EPR data are being integrated, too, correct?

**Comment by xu-hao** [comment]
I was working on it but hit a roadblock because the crosswalk file didn't match the id in the epr file. I sent and email and we discussed the issue and I think Emily is working on generating a new crosswalk file.

**Comment by karafecho** [comment]
Yes, I recall that discussion, but I thought you were moving ahead with the roughly 600 matches while you waited for the new crosswalk from @empfff. If it makes more sense to wait, that's fine; I just want to make sure we're all on the same page.

**Comment by xu-hao** [comment]
I don't have the 600 matches. If @empfff  can send me the file, I can move ahead.

**Comment by empfff** [comment]
I just got back from AMIA today. I will try to get you the matches tomorrow.

**Comment by karafecho** [comment]
Great! I now see my confusion: I forgot that it was Emily (not Hao) who identified the roughly 600 matches.

**Comment by karafecho** [comment]
@jameschump @xu-hao : I'm wondering if the remaining issues with the 2010-2016 patient files have been resolved. Thanks!

**Comment by xu-hao** [comment]
@karafecho Currently 2010 and 2011 data are no longer available on rockfish. I believe @empfff  will put it back on there, but according to our meeting it was not the first task she would be tackling. I tried running the pipeline on 2014 to 2016 data and I've sent @jameschump  emails on issues with 2014 and 2015 data. Currently, the pipeline on 2016 data is running.

**Comment by empfff** [comment]
@xu-hao can you document these issues you're finding in this thread? it will be easier for me and James to work together to solve things if I can understand what the issues are as well. It is not clear to me if these are minor tweaks, or major issues.

Regarding the 2010 and 201 data--I can recreate them, but as I mentioned on the call, my thought was that those would not be as high priority of a task seeing as there are already finished ICEES tables for those years.

**Comment by xu-hao** [comment]
@empfff 

2014: 
Observation_Labs/5424000.json seems to be having only part of a json object. It looked like it was not a complete json object.

2015:
MedicationRequestion some records in the files doesn't have medicationCodeableConcept, but have medicationReference, which differs from data from other years. In other years, the records have medicationCodeableConcept field but not medicationReference.

We have icees tables for 2010 but not 2011. With 2010 we only have the old columns. I could just use the old table and join it with additional columns that are added since it was created. But for reproducibility, I'd like to rerun the pipeline on 2010 data.

**Comment by karafecho** [comment]
I agree that we should try to maintain all documentation here, just so that we're all on the same page and so things don't slip through, especially as we (hopefully) work toward the March 2020 demonstration project. 

WRT 2010 and 2011, how hard/time-consuming is it to generate new tables using CAMP FHIR? I don't have a good feel for this, but I probably should know for future reference. Thanks!

**Comment by empfff** [comment]
@karafecho it's not that time consuming to do it for data that's available in PCORnet, but running the pipeline in i2b2 to grab the older stuff is pretty challenging/time consuming. Pulling data for 2017 would be a snap--re-pulling 2010, less so.

**Comment by karafecho** [comment]
@empfff : I see. Thanks for the explanation. 

So, we have two versions of 2010 data available via the ICEES OpenAPI, although they are out of date in terms of the number of available feature variables. A version of 2011 data WAS available via the ICEES OpenAPI, but I asked Hao to remove the data because there were issues with the exposures data for that year.  We could focus on 2012-2016 initially; then re-run the pipeline for 2011; and then re-run the pipeline for 2010 (to expose v3 of 2010 data). Does this plan make sense to folks?

In the meantime, I can submit an IRB modification so that we can access 2017-2019 data.

**Comment by karafecho** [comment]
@empfff @jameschump @xu-hao  : We are submitting a no-cost extension for Green Team's Translator award. I am adding the following specific activity: Expanded ICEES Integrated Feature Tables.
We will work to expand the existing the ICEES integrated feature tables to include additional variables, based on new use cases, and additional years of data, beyond the most recent available calendar year of 2016. This work will be completed as part of Feasibility Milestones 1, 5, and 13.

**Comment by xu-hao** [comment]
@jameschump @empfff  I noticed that in the 2016 data there is no display field under MedicationRequestion's medicationCodeableConcept field, I've been using that field to match medication name for generating icees features, is it possible to add that field or is there a preferred way to do this?

**Comment by karafecho** [comment]
Julie Hayes will be submitting the annual IRB renewal in January 2020. At that time, a protocol modification will be requested to: (1) remove individuals who do not require access to the identified patient data; and (2) extend the study to include calendar years 2017, 2018, and 2019.

**Comment by karafecho** [comment]
Update from 12.11.19:

1. @jameschump is re-running CAMP FHIR to generate new year 2015 patient data and fixing errors with year 2014 patient data. Also regenerating 2011 patient data, which was deleted from Rockfish and thus no longer accessible.
2. @xu-hao will use the RxNorm IDs available in [this spreadsheet](https://github.com/NCATS-Tangerine/icees-api/blob/master/docs/ICEES_Identifiers_v7%2006.03.19.xlsx) to identify medications for years 2014, 2015, and 2016. Hao is also re-running FHIR PIT to generate new 2012 and 2013 tables.
3. @mmersmann will reach out to Jim Balhoff re UMLS licensing; team may reach out directly to UMLS for clarification on licensing.

**Comment by jameschump** [comment]
I regenerated 2015 medication data and pushed onto rockfish. Let me know if you see any issues.

**Comment by xu-hao** [comment]
Thanks!

**Comment by jameschump** [comment]
Data for 2014 Observation Labs has been regenerated.

________________________________

**Comment by karafecho** [comment]
@jameschump : Where do things stand with the 2011 patient data?

**Comment by karafecho** [comment]
@xu-hao : Please confirm that the RxNorm IDs available in the spreadsheet linked above are sufficient, at least as a temporary fix, for the 2016 patient data.

**Comment by karafecho** [comment]
Hao confirmed that the RxNorm identifiers that I provided in the spreadsheet above are adequate for the 2015 and 2016 patient files.

**Comment by karafecho** [comment]
@empfff @xu-hao : Perhaps one of you can remind me of the status of data on pulmonary lung function tests? Mike Knowles and Margaret Leigh are interested in capturing this, as is Dave. I remember investigating this with Emily, but I can't recall the outcome.

**Comment by empfff** [comment]
we should be able to get these--they're not in the current data pulls, though, so we would have to build the capability to handle them. not a huge ask.

**Comment by karafecho** [comment]
Great, @empfff. Let's prioritize this after we (1) resolving the hashing issues and (2) work out the remaining kinks in the 2010-2016 patient data. Does this seem like a reasonable plan?

**Comment by empfff** [comment]
We are unaware of kinks in the 2010-2016 data other than needing to recreate the raw files for 2010-2011, which we are working through now. Are there other things that james and I need to handle?

**Comment by karafecho** [comment]
As of the moment, the only issues relate to recreating new 2010 and 2011 patient files. Thanks for checking.

**Comment by karafecho** [comment]
From Emily, new hash matching results, 1/10/2020:

UNCHCS denominator: 2,770,607 patients

NIEHS EPR denominator: 19,388 participants

Matched: 7,233 people (37% of all EPR participants)

**Comment by karafecho** [comment]
From Emily: The “identifier” I have for the NIEHS participants is a hash. Because we redid the hashing effort, this hash will not be the same NIEHS hash that Hao currently has in his possession. This means that NIEHS will need to either re-run their data for Hao with the new hash swapped in as the patient identifier, OR will need to produce a hash crosswalk for Hao so he can translate between the old and new hashes. When NIEHS does this, the hash I am going to give to Hao as an NIEHS “identifier” is what is labelled “Hash 1” in her dataset.

**Comment by karafecho** [comment]
NC TraCS update, 1/13/20:

2010 and 2011 patient data expected next week. Using PCORnet, but adding i2b2 meds to the data.

**Comment by karafecho** [comment]
From Emily, 1/14/20:

Crosswalk is up on Rockfish, in /opt/RENCI/output/FHIR. Filename is UNC_NIEHS_XWalk_for_Hao.csv. The UNC identifier should match the patient IDs you have the in the FHIR files. The hash is what will match with NIEHS.

**Comment by karafecho** [comment]
Update from Kara on EPR asthma cohort data, 01.16.20:

N= 4129 total, all with SNP data

N= 2709 with HE_COMPLETION_DATE
Of those 2709, 4 participants have HE_COMPLETION_DATE in 2014, and 5 have HE_COMPLETION_DATE in 2018, remainder are all 2012 and 2013

N = 2637 with HE_COMPLETION_DATE and D28_Asthma = 0,1
Of those 2637, 1 participant has HE_COMPLETION_DATE in 2014, remainder are all 2012 and 2013
If 5% overlap with UNC ICEES asthma, then N = 132
If 37% overlap with UNC ICEES asthma, then N = 975

Of those 2637, 928 have HE_COMPLETION_DATE and D28_Asthma = 1
If 5% overlap with UNC ICEES asthma, then N = 46
If 37% overlap with UNC ICEES asthma, then N = 343

N = 2593 with D28_Asthma = 0,1 and TLR4_DIST_1X

**Comment by karafecho** [comment]
Additional update from Kara, 01.16.20:

Hao has lat/longs and addresses for all 2709 participants with HE_COMPLETION_Date, so we can integrate the exposures data for the 2705 participants with HE_COMPLETION_DATE in 2012 or 2013, excluding the 4 participants with HE_COMPLETION_DATE in 2014. 

We will use use the calendar year prior to HE_COMPLETION_DATE to determine airborne pollutant exposure estimates, using the same calculations for AVG and MAX exposure that we currently use for the ICEES integrated feature tables, but expanding from PM2.5 and ozone to include the eight additional airborne pollutant exposure estimates that we now have. 

In other words, we'll calculate one-year exposures over the year prior to survey completion. So, if someone completed the survey on 7/1/2012, then we'll calculate exposures from 7/1/2011 - 7/1/2012 (or 7/2/2011 - 7/1/2012). 

For the ACS data, we will use the 2012-2016 estimates.

 For the roadway data, we do not really have a choice, as we only have 2016 data.

WRT to column headers for the EPR data, we'll create two sets: one for the UNC data and one for the EPR data. 

We'll integrate the UNC and EPR data over all available years to date, i.e., 2010-2016.

@xu-hao : does this plan seem correct to you?

**Comment by xu-hao** [comment]
yes, your plan seems correct

**Comment by karafecho** [comment]
@charles.schmitt@nih.gov : Any updates on the S3 hash values for the EPR asthma cohort? We cannot move forward with this effort until we receive the new hash values.

**Comment by karafecho** [comment]
@jameschump @empfff : Any updates on the 2010 and 2011 patient data? Thanks!

**Comment by empfff** [comment]
Still aiming for this week as we discussed, but we have a dependency on getting the legacy medication data loaded. We are working on it.

**Comment by xu-hao** [comment]
@jameschump @empfff I found this following issue in the 2015 data

files under Condition:
code -> coding only has system field but not code field
example: 100000.json

**Comment by jameschump** [comment]
2010 and 2011 data are now available on rockfish.

**Comment by karafecho** [comment]
Thanks, @jameschump!

Has the minor issue with the 2015 data been resolved?

**Comment by jameschump** [comment]
I haven't had a chance to look at it yet but plan to today.
